Add the memory swappiness tuning option to docker.

ktraghavendra opened the pull request:
Memory swappiness option takes 0-100, and helps to tune swappiness
behavior per container.

Signed-off-by: Raghavendra K T raghavendra.kt@linux.vnet.ibm.com

ktraghavendra commented:
Is that a false build failure (experimental) ?

LK4D4 commented:
@ktraghavendra Rebase is needed.

ktraghavendra commented:
@LK4D4 I have rebased the patch, but planning to introduce -1 value to identify the situation when
no mem-swappiness value was passed as argument to docker UI. From this want to address https://github.com/docker/libcontainer/issues/639

ktraghavendra commented:
@LK4D4 @estesp @dqminh 
I have created https://github.com/docker/libcontainer/pull/654 which means that we treat -1 as default value (nobody has set memswappiness from UI or it is 60 ). and we skip the enforcement of memory swappiness.
I have accordingly updated docker UI patch too

ktraghavendra commented:
@LK4D4 just FYI .. 
I corrected go error format.
this i safer to be pulled even without libcontainer counterpart of https://github.com/docker/libcontainer/issues/639

LK4D4 commented on runconfig/parse.go:
```diff
@@ -59,6 +59,7 @@ func Parse(cmd *flag.FlagSet, args []string) (*Config, *HostConfig, *flag.FlagSe
 		flTty             = cmd.Bool([]string{"t", "-tty"}, false, "Allocate a pseudo-TTY")
 		flOomKillDisable  = cmd.Bool([]string{"-oom-kill-disable"}, false, "Disable OOM Killer")
+		flSwappiness      = cmd.Int64([]string{"-mem-swappiness"}, 60, "Tuning container memory swappiness")
 		flContainerIDFile = cmd.String([]string{"#cidfile", "-cidfile"}, "", "Write the container ID to the file")
```
I suggest use `memory-swappiness` for consistency with `memory-swap`

ktraghavendra replied:
@LK4D4. Thanks.. done. (to memory-swappiness)

ktraghavendra commented:
@LK4D4 , do you think we have some pending things for merging the patch? please do let me know

LK4D4 commented:
LGTM

cpuguy83 commented on docs/reference/commandline/run.md:
```diff
@@ -58,6 +58,7 @@ weight=1
       --read-only=false          Mount the container's root filesystem as read only
       --restart="no"             Restart policy (no, on-failure[:max-retry], always)
       --rm=false                 Automatically remove the container when it exits
+      --memory-swappiness=[]     Tuning container memory swappiness (accepted value 0-100)
```
This should be `--memory-swapiness=60`

cpuguy83 commented on docs/reference/commandline/create.md:
```diff
@@ -57,6 +57,7 @@ Creates a new container.
       --privileged=false         Give extended privileges to this container
       --read-only=false          Mount the container's root filesystem as read only
       --restart="no"             Restart policy (no, on-failure[:max-retry], always)
+      --memory-swappiness=[]     Tuning container memory swappiness (accepted value 0-100)
```
This should be --memory-swapiness=60

cpuguy83 commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -61,6 +61,14 @@ func (s *DockerSuite) TestRunWithoutMemoryswapLimit(c *check.C) {
 }
 
+func (s *DockerSuite) TestRunWithSwappiness(c *check.C) {
+	runCmd := exec.Command(dockerBinary, "run", "--memory-swappiness", "0", "busybox", "true")
```
some extra tests here showing that it can't be set less than 0 or greater than 100 would be fantastic.
Also probably demonstrating the default value is set to 60 would be great (to make sure this doesn't change by accident).

ktraghavendra replied:
Agree. Adding invalid value check test. 'll try on how default value test too.

cpuguy83 replied:
You can check the hostconfig struct (using the inspect helper) after the container is created.

ktraghavendra replied:
@cpuguy83 Thanks for hints.. will look at that

ktraghavendra replied:
@cpuguy83 , @lizf-os @hqhq with the current implementation we do not  change the existing value of container inherited by parent.
Which one is better according to you?

1. let the container inherit the value from parent (I 'll have to change the documentation too to reflect that)

2. let us override default value to 60
( This would not work with old kernels where we are not allowed to write swappiness value)

hqhq replied:
If user didn't use `--memory-swappiness`, of course we should inherit value from parent. Current logic looks good to me (use default value -1 which will be ignored by libcontainer).

ktraghavendra replied:
@hqhq 
Okay, To summarize can we have consensus on: 
user specifies --memory-swappiness  <value> => use value
user specifies --memory-swappiness => use deafult = 60
user does not specify --memory-swappiness => inherit the value from parent

so code should be something like below ?
var parsedSwappiness int64
        var flSwappiness int64
        flSwappiness = -1

        if *flSwappinessStr != "" {
                parsedSwappiness, err = strconv.ParseInt(*flSwappinessStr, 10, 64)
                if err != nil || parsedSwappiness < 0 || parsedSwappiness > 100 {
                        return nil, nil, cmd, fmt.Errorf("invalid value:%s. valid memory swappiness range is 0-100", *flSwappinessStr)
                }
                flSwappiness = parsedSwappiness
        } else {
                flSwappiness = 60
        }

ktraghavendra replied:
I was too fast regarding the code. will do some more testing to arrive at the consensus

ktraghavendra replied:
1. user specifies --memory-swappiness value => use value
2. user specifies --memory-swappiness =>  no value specified use deafult = 60
3. user does not specify --memory-swappiness => inherit the value from parent
I could not differentiate between (2) and (3) ..
@hqhq @lizf-os @cpuguy83 please let me know if you have any suggestions
Note that current behaviour is unless we override it uses whatever is inherited from parent..
so except documentation change rest of the code is in sync with what we expect.

lizf-os commented:
You are assuming the default value of swappiness of a cgroup is 60, which is not true. The init value is inherited from the parent.

hqhq commented:
@lizf-os I think the stratagem here is when you use `--MemorySwappiness` without any following value, we'll set the default value as 60, if we don't use `--MemorySwappiness`, the default swappiness value for the container would be inherited from the parent, which I think is reasonable, WDYT?

lizf-os commented:
@hqhq If I understand the code correctly, when I run "docker run --memory-swappiness=60", docker won't write to memory.swappiness at all. Now if the parent cgroup's swappiness is 50, this command won't set memory.swappiness to 60.

hqhq commented:
Yes, you are right, I didn't review in details, now I see that part is suspicious, maybe we can comment in the code.

hqhq commented on runconfig/parse.go:
```diff
@@ -187,6 +188,13 @@ func Parse(cmd *flag.FlagSet, args []string) (*Config, *HostConfig, *flag.FlagSe
+	// This is to ensure that we do not try to override swappiness value for default case
+	if *flSwappiness == 60 {
+		*flSwappiness = -1
+	}
```
By default, swappiness will be inherited from parent, so if parent is 50, and you set `--memory-swappiness=60` won't set swappiness to 60 as user may expected.

ktraghavendra replied:
@hqhq @lizf-os Yes Agree.. there is a gap in handling the special case you mentioned. I have been thinking over how to

case1: handle if somebody tries to set swappiness = 60 explicitly
case2: nobody asks us to configure swappiness
In case1 we should try to write the value.  (case2 implicitly swappiness value should be set to -1)
so that we ignore writing any value.

hqhq replied:
You don't need to check any specific value, as I said in thread comment, you already treated `--memory-swappiness` 's default value as 60, so if user didn't assign any value, the default would be 60, if user didn't use option `--memory-swappiness`, the value would be inherited from parent which is done by cgroup itself. Which makes sense to me.

So you don't need to anything here, just remove this block.

ktraghavendra replied:
@hqhq  Removing that block makes sense.
I will check once again if that would result in errors while creating containers for older 14.04 kernels
with use_hierarchy=1.

hqhq replied:
Yes, there would be problem if we using it on kernel older than 3.16 with use_hierarchy=1, but I remember there is another PR handling this, no?
Probably we should ignore any value set to swappiness in this case, or just throw error out. If you want to handle specific kernel version, might pick up this commit: https://github.com/hqhq/docker/commit/16a0e0ec51c092eb28c3f4e96ff70afbfb879b66

ktraghavendra replied:
@hqhq Yes there is a different PR. Now I am able to get the patch to handle swappiness correctly
in Docker UI. pushing.

ktraghavendra replied:
@lizf-os Hope latest logic looks good to you.

hqhq commented on runconfig/parse.go:
```diff
@@ -188,6 +189,19 @@ func Parse(cmd *flag.FlagSet, args []string) (*Config, *HostConfig, *flag.FlagSe
+		flSwappiness = parsedSwappiness
+	} else {
+		flSwappiness = -1
+	}
```
There are something wrong with my former comments. The default value is for you didn't use the option, not for you use an option without any following values.

So here you just want the default swappiness to be -1, I think you can set:
```
flSwappinessStr   = cmd.Int64([]string{"-memory-swappiness"}, -1, "Tuning container memory swappiness (0 to 100)")
```
It'll be much simpler.

ktraghavendra replied:
@hqhq But the problem here was to differentiate whether use explicitly asked for an invalid value of 
-1 or he did not specify anything so that we have the by default -1 set.. or am I missed something?

hqhq replied:
I think only bool options can be used without specify any values, so here we only need to deal with specified value or didn't use this option, which will just use the default value we specified in `cmd.Int64()`.

ktraghavendra replied:
@hqhq Thanks for confirming .. Okay then we are left only with whether 
1. we use 60 as default value
2. we use parent inherited swappiness value as default value..

hqhq replied:
I would prefer use parent's value by default.

ktraghavendra replied:
@hqhq Yes that is the current status of code.

ktraghavendra commented:
@hqhq @lizf-os @cpuguy83 @LK4D4 ,
In summary:
current status: 
we are at: 
case 1.we have the patch that by default makes swappiness inherited by parent

case 2: swappiness by default = 60

if we have consensus on  case1:
I ll update the patch to remove 60 as default value stated in documents

if we have consensus on case2:
I 'll make default as 60
add a cli run test case to check default as 60 using inspect as suggested by @cpuguy83

ktraghavendra commented:
@hqhq @lizf-os @cpuguy83 @LK4D4 
now I have the code which conforms to having swappiness inherited by parent.

cpuguy83 commented:
Do we need to check these when setting swappiness?
1. memswap is not disabled for the container
2. that swapiness is supported?

ktraghavendra commented:
@cpuguy83 
1. memswap is not disabled for the container
memswap disabled is a misnomer perhaps. setting memswap=-1 does not disable actual swapping.
basically memswap=-1 => unlimited swapping. (otherwise memory.memsw.limit_in_bytes get set explicitly to twice memory limit or as specified).
So no need to check for this case.

2. that swapiness is supported?
a.When user does not ask for swappiness, writing to memory.swappiness does not happen so no need to check for this case.

b. when user asks for swappiness writing swappiness will get error and it is expected.

So in summary, we do not need any check.

ktraghavendra commented:
@cpuguy83 @lizf-os , could please let me know if this looks good for merging?
(I think @hqhq was latest code and we have  LKDM from @LK4D4)

estesp commented on docs/reference/api/docker_remote_api_v1.20.md:
```diff
@@ -208,6 +209,7 @@ Json Parameters:
 -   **CpusetCpus** - String value containing the `cgroups CpusetCpus` to use.
 -   **CpusetMems** - Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
 -   **BlkioWeight** - Block IO weight (relative weight) accepts a weight value between 10 and 1000.
+-   **MemorySwappiness** - Integer Value, Tuning Container memory swappiness. Accepts value between 0 and 100.
```
the 2nd, 3rd, and 4th words should not be capitalized ("**Integer value, tuning container's memory swappiness**"); I also think "value" should be plural: ("**Accepts values between 0 and 100.**")

ktraghavendra replied:
@estesp Thanks for comments
1. changed to: Integer value, tuning container's memory swappiness
2. regarding Accepts values between 0 and 100: how about changing to
Accepts a value between 0 and 100, since we should be specifying only one value in that range at any point of time.

estesp replied:
"accepts a value" is good with me as well, thanks @ktraghavendra

estesp commented on docs/reference/commandline/create.md:
```diff
@@ -57,6 +57,7 @@ Creates a new container.
       --privileged=false         Give extended privileges to this container
       --read-only=false          Mount the container's root filesystem as read only
       --restart="no"             Restart policy (no, on-failure[:max-retry], always)
+      --memory-swappiness=""     Tuning container memory swappiness (accepted value 0-100)
```
"accepted value**s**" <-- plural

ktraghavendra replied:
changing to : Tuning container's memory swappiness (accepted values 0-100)

estesp commented on docs/reference/commandline/run.md:
```diff
@@ -58,6 +58,7 @@ weight=1
       --read-only=false          Mount the container's root filesystem as read only
       --restart="no"             Restart policy (no, on-failure[:max-retry], always)
       --rm=false                 Automatically remove the container when it exits
+      --memory-swappiness=""     Tuning container memory swappiness (accepted value 0-100)
```
same comment re: value vs. values

ktraghavendra replied:
yes, as above-> Tuning container's memory swappiness (accepted values 0-100)

estesp commented:
I think this is in good shape, but @moxiegirl or other docs writers should review the documentation

code LGTM

hqhq commented:
Agreed with @cpuguy83 If we don't have swap memory support for cgroup, there is no meaning we set swappiness, should show a warning to users. Although we'll get some errors anyway, we need to keep these message unified, like all other cgroup options.

ktraghavendra commented:
@hqhq If you prefer to show warning too like we do for other cgroups options.. yes, I 'll add that

ktraghavendra commented:
@hqhq Please let me know if you are fine with the warning added (similar to OomKillDisable etc of cgroup config)

ktraghavendra commented:
@LK4D4 @cpuguy83 @lizf-os @hqhq @estesp @moxiegirl  Seems like Janky status in incorrect. and this can be pulled if there are no further comments.

cpuguy83 commented:
@ktraghavendra This still needs to finish up docs review.

moxiegirl commented:
checking

moxiegirl commented on docs/reference/run.md:
```diff
@@ -611,6 +612,17 @@ The following example, illustrates a dangerous way to use the flag:
 The container has unlimited memory which can cause the host to run out memory
 and require killing system processes to free memory.
 
+### Swappiness constraint
```
@ktraghavendra thank you for the contribution.  There is some awkward phrasing so, the content is hard to follow.  How about this?

---> https://gist.github.com/moxiegirl/6a0d6e72767b1c10ad29

### Swappiness constraint

By default, a container's kernel can swap out a percentage of anonymous pages.
To set this percentage for a container, specify a `--memory-swappiness` value between 0 and 100. 
A value of 0 turns off anonymous page swapping.  A value of 100 sets all anonymous pages as swappable. For example, you can set:

    $ docker run -ti --memory-swappiness=0 ubuntu:14.04 /bin/bash

Setting the  `--memory-swappiness` option is helpful when you want to retain the container's working set and to avoid swapping performance penalties.

moxiegirl commented on docs/reference/run.md:
```diff
@@ -514,6 +514,7 @@ container:
     --cpu-quota=0: Limit the CPU CFS (Completely Fair Scheduler) quota
     --blkio-weight=0: Block IO weight (relative weight) accepts a weight value between 10 and 1000.
     --oom-kill-disable=true|false: Whether to disable OOM Killer for the container or not.
+    --memory-swappiness="": Tuning the swappiness behavior of container accepts value between 0 and 100.
```
Tune a container's swappiness behavior. Accepts an integer between 0 to 100.

moxiegirl commented on docs/reference/commandline/run.md:
```diff
@@ -58,6 +58,7 @@ weight=1
       --read-only=false          Mount the container's root filesystem as read only
       --restart="no"             Restart policy (no, on-failure[:max-retry], always)
       --rm=false                 Automatically remove the container when it exits
+      --memory-swappiness=""     Tuning container's memory swappiness (accepted values 0-100)
```
Tune a container's swappiness behavior. Accepts an integer between 0 to 100.

moxiegirl commented on docs/reference/commandline/create.md:
```diff
@@ -57,6 +57,7 @@ Creates a new container.
       --privileged=false         Give extended privileges to this container
       --read-only=false          Mount the container's root filesystem as read only
       --restart="no"             Restart policy (no, on-failure[:max-retry], always)
+      --memory-swappiness=""     Tuning container's memory swappiness (accepted values 0-100)
```
Tune a container's memory swappiness behavior. Accepts an integer between 0 to 100.

moxiegirl commented on docs/reference/api/docker_remote_api_v1.20.md:
```diff
@@ -208,6 +209,7 @@ Json Parameters:
 -   **CpusetCpus** - String value containing the `cgroups CpusetCpus` to use.
 -   **CpusetMems** - Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
 -   **BlkioWeight** - Block IO weight (relative weight) accepts a weight value between 10 and 1000.
+-   **MemorySwappiness** - Integer value, tuning container's memory swappiness. Accepts a value between 0 and 100.
```
Tune a container's memory swappiness behavior. Accepts an integer between 0 to 100.

moxiegirl commented on man/docker-create.1.md:
```diff
@@ -225,6 +226,9 @@ This value should always larger than **-m**, so you should always use this with
    Security Options
 
+**--memory-swappiness**=*60*
+   Tuning the memory swappiness of container. The value in the range of 0-100 is accepted.
```
Tune a container's memory swappiness behavior. Accepts an integer between 0 to 100.

moxiegirl commented on man/docker-run.1.md:
```diff
@@ -371,6 +372,9 @@ its root filesystem mounted as read only prohibiting any writes.
    Proxy received signals to the process (non-TTY mode only). SIGCHLD, SIGSTOP, and SIGKILL are not proxied. The default is *true*.
 
+**--memory-swappiness**=*60*
+   Tuning the memory swappiness of container. The value in the range of 0-100 is accepted.
```
Tune a container's memory swappiness behavior. Accepts an integer between 0 to 100.

ktraghavendra committed:
Add the memory swappiness tuning option to docker.

Memory swappiness option takes 0-100, and helps to tune swappiness
behavior per container.
For example, When a lower value of swappiness is chosen
the container will see minimum major faults. When no value is
specified for memory-swappiness in docker UI, it is inherited from
parent cgroup. (generally 60 unless it is changed).

Signed-off-by: Raghavendra K T <raghavendra.kt@linux.vnet.ibm.com>

ktraghavendra commented:
@moxiegirl  Thanks for the comments. Have incorporated them all.

moxiegirl commented:
Thanks @ktraghavendra! Ping @thaJeztah

thaJeztah commented on docs/reference/api/docker_remote_api_v1.20.md:
```diff
@@ -208,6 +209,7 @@ Json Parameters:
 -   **CpusetCpus** - String value containing the `cgroups CpusetCpus` to use.
 -   **CpusetMems** - Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
 -   **BlkioWeight** - Block IO weight (relative weight) accepts a weight value between 10 and 1000.
+-   **MemorySwappiness** - Tune a container's memory swappiness behavior. Accepts an integer between 0 to 100.
```
can you change "between 0 to 100." to "between 0 and 100." ?

thaJeztah commented on docs/reference/commandline/create.md:
```diff
@@ -57,6 +57,7 @@ Creates a new container.
       --privileged=false         Give extended privileges to this container
       --read-only=false          Mount the container's root filesystem as read only
       --restart="no"             Restart policy (no, on-failure[:max-retry], always)
+      --memory-swappiness=""     Tune a container's memory swappiness behavior. Accepts an integer between 0 to 100.
```
"between 0 to 100."  ->  "between 0 and 100."

thaJeztah commented on docs/reference/commandline/run.md:
```diff
@@ -58,6 +58,7 @@ weight=1
       --read-only=false          Mount the container's root filesystem as read only
       --restart="no"             Restart policy (no, on-failure[:max-retry], always)
       --rm=false                 Automatically remove the container when it exits
+      --memory-swappiness=""     Tune a container's memory swappiness behavior. Accepts an integer between 0 to 100.
```
"between 0 to 100."  ->  "between 0 and 100."

thaJeztah commented on docs/reference/run.md:
```diff
@@ -514,6 +514,7 @@ container:
     --cpu-quota=0: Limit the CPU CFS (Completely Fair Scheduler) quota
     --blkio-weight=0: Block IO weight (relative weight) accepts a weight value between 10 and 1000.
     --oom-kill-disable=true|false: Whether to disable OOM Killer for the container or not.
+    --memory-swappiness="": Tune a container's memory swappiness behavior. Accepts an integer between 0 to 100.
```
"between 0 to 100."  ->  "between 0 and 100."

thaJeztah commented on man/docker-create.1.md:
```diff
@@ -225,6 +226,9 @@ This value should always larger than **-m**, so you should always use this with
    Security Options
 
+**--memory-swappiness**=""
+   Tune a container's memory swappiness behavior. Accepts an integer between 0 to 100.
```
"between 0 to 100."  ->  "between 0 and 100."

thaJeztah commented on man/docker-run.1.md:
```diff
@@ -371,6 +372,9 @@ its root filesystem mounted as read only prohibiting any writes.
    Proxy received signals to the process (non-TTY mode only). SIGCHLD, SIGSTOP, and SIGKILL are not proxied. The default is *true*.
 
+**--memory-swappiness**=""
+   Tune a container's memory swappiness behavior. Accepts an integer between 0 to 100.
```
"between 0 to 100."  ->  "between 0 and 100."

thaJeztah commented on docs/reference/run.md:
```diff
@@ -611,6 +612,20 @@ The following example, illustrates a dangerous way to use the flag:
+### Swappiness constraint
+
+By default, a container's kernel can swap out a percentage of anonymous pages.
+To set this percentage for a container, specify a --memory-swappiness value
```
Add back-ticks around `--memory-swappiness`

thaJeztah commented on docs/reference/run.md:
```diff
@@ -611,6 +612,20 @@ The following example, illustrates a dangerous way to use the flag:
+
+    $ docker run -ti --memory-swappiness=0 ubuntu:14.04 /bin/bash
+
+Setting the --memory-swappiness option is helpful when you want to retain the
```
Same here (back-ticks)

thaJeztah commented:
Thanks for your contribution, @ktraghavendra. I added some notes above

ktraghavendra commented:
@thaJeztah Thanks for the comments. Have incorporated your comments.

thaJeztah commented:
docs LGTM. Thanks for the fast turnaround, @ktraghavendra 


ping @fredlf @jamtur01 @moxiegirl @SvenDowideit

jamtur01 commented:
Docs LGTM

thaJeztah commented:
ping @LK4D4 @cpuguy83 @estesp could you do a final check if this LGTY? (I see some code changed after LGTM's, so want to be sure)

estesp commented:
yes, just looked at the code again; still LGTM and checked with @cpuguy83 as well

thaJeztah commented:
Thanks @estesp!

vieux commented:
@ktraghavendra @estesp accepted values are from 0 to 100 but the default value is -1 ? Seems odd to me, it forces every client out there to set a default value to -1 instead of simply not filling out the field, is this correct ?

 We have an issue in https://github.com/docker/swarm/issues/1411

It might come from compose

ktraghavendra commented:
@vieux -1 indicates that the value of memory swappiness should be inherited from parent and any explicit setting of swappiness should not be done. 0-100 are valid values of swappiness so setting
0 by default is also incorrect.
[ normally memory swappiness is 60 but it could be that parent cgroup value is overriden in which case setting value=60 in the child cgroup is not correct ]
for the above reasons, default value concept was little tricky to handle.
